packages/styled-components/src/types.ts

```diff
@@ -92,7 +178,21 @@ export type Interpolation<Props extends object> =
   | Keyframes
   | StyledComponentBrand
   | RuleSet<Props>
-  | Interpolation<Props>[];
+  | Interpolation<Props>[]
+  | InterpolationFunction<Props>
+  | FlattenInterpolation<Props>;
+
+export type InterpolationFunction<P extends object> = (props: P) => Interpolation<P>;
```

Comment:
`InterpolationFunction` and other types here looks like they come from [the Definitely Typed definitions for v5](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/styled-components/index.d.ts#L106).

`InterpolationFunction` from v5 can be replaced with `StyleFunction` from v6. 

Maybe we can either write a migration guide from v5 to v6, or consider just aliasing some of the types that are 1 to 1 replacements? 

I don't think we should copy all of the types over from Definitely Typed because there are new, similar types defined in v6. It might get really confusing to maintain two very similar copies of the same type but with different names.